### DISCUSSION TITLE
Windows container networking is not stable

### DISCUSSION THREAD

**Discussion by jvnvenu** [rancher/rke2#6100] [Q&A]
**Environmental Info:**
RKE2 Version: 
1.29.2

Node(s) CPU architecture, OS, and Version: 
4 Nodes, 3 Linux (Redhat 8), 1 windows (windows server 2022)

Cluster Configuration:
3 linux servers, 1 windows agent

**Describe the bug:**
I have multiple deployments which have init and normal containers in each one. The init container will call the some Kube APIs to get resources information also update some annotations in the same deployment. Also each deployment has a service running on and one service can communicate with another. The Linux init container and communication between linux service works fine. 
But windows init container when it is calling Kuber api get resource information it is getting connection closed by remote host issue. Then it crashing and starting again. Sometime it is starting without error. Sometime 2 or 3 times get this error then starts fine.
Same way from normal containers if it is trying to access a linux service it is getting timeout issue. After 1 or 2 retry it is successful.
It looks to me the windows rke doesn't have stable kubenetes connectivity with windows Nodes.
In linux we are using calico.

**Steps To Reproduce:**
Run a application to update annotations from init container for the same running deployment.

**Expected behavior:**
Should not see any communication failures in windows containers

**Actual behavior:**
Lot of communication failures in windows containers

**Additional context / logs:**

**Comment by manuelbuil** [comment]
Do you have network policies? `kubectl get netpol -A`.  There is a known issue in the windows tool that calico uses to program network policies in Windows: https://docs.tigera.io/calico/latest/getting-started/kubernetes/windows-calico/limitations#pod-to-pod-connections-are-dropped-with-tcp-reset-packets

**Comment by jvnvenu** [comment]
We don't have any customized polices. I ran the command the results are given below. Looks to me these are default. Please share your thoughts.

NAMESPACE                       - NAME                                 -  POD-SELECTOR    
cattle-fleet-local-system       - default-allow-all                    -  <none>          
cattle-provisioning-capi-system - default-allow-all                    -  <none>          
kube-system                     - rancher-monitoring-coredns-allow-all -  k8s-app=kube-dns

**Comment by manuelbuil** [comment]
It does not matter if they are default. Any creation of a pod matching those policies or any deletion of a pod matching those policies will reset the ACL object in Windows. As a consequence, any TCP Session established by any pod included in that ACL will TCP-RST

**Comment by manuelbuil** [comment]
Give `cni: flannel` a try and see if you are still experiencing the same instability. That will help you understand if this is the issue. Flannel does not implement any network policy, so this problem is not affecting it

**Comment by jvnvenu** [comment]
But the above mentioned polices are not selecting the windows pods. It is selecting 1 pod each policy and those are running under linux nodes.
Can flannel be used as a production setup instead of calico with linux and windows nodes same time. Because as per RKE2 documentation 

**Windows Air-Gap Install**
**Windows Support is currently Experimental as of v1.21.3+rke2r1 Windows Support requires choosing Calico as the CNI for the RKE2 cluster**

**Comment by brandond** [comment]
Note that the docs say 
> Some ingress or egress policy that applies to a pod contains selectors and the set of endpoints that those selectors match changes.

Note that the endpoint that the Windows pod communicates with does not ALSO have to be on Windows; the Windows pod may experience a disruption if a pod on a completely different node is recreated and the endpoint IP changes. This is a defect in Windows's HNS subsystem, not in canal, containerd, kubernetes, or RKE2.

Please try flannel as @manuelbuil suggested.

**Comment by manuelbuil** [comment]
> But the above mentioned polices are not selecting the windows pods. It is selecting 1 pod each policy and those are running under linux nodes. Can flannel be used as a production setup instead of calico with linux and windows nodes same time. Because as per RKE2 documentation
> 
> **Windows Air-Gap Install** **Windows Support is currently Experimental as of v1.21.3+rke2r1 Windows Support requires choosing Calico as the CNI for the RKE2 cluster**

Due to the network policy problem, we introduced Flannel recently. I missed that line in the docs! Thanks, I'll change it :)

Flannel can be used as production, yes

**Comment by manuelbuil** [comment]
https://github.com/rancher/rke2-docs/pull/220

**Comment by jvnvenu** [comment]
After changed the cni to flannel it looks promising. Thanks for the help

**Comment by jvnvenu** [comment]
sometime flanneld.exe is not starting and all the calls to KubeAPIs are failing when i kill all containerd processes and restart rke2 process then flanneld.exe is starting. It is consistently happening even on machine restart

**Reply by manuelbuil** [reply]
Probably you are hitting an issue that happens in windows when there is only one interface. When we create the vxlan network, the physical interface bound to that vxlan network disappears for some seconds (sometimes up to 30s). All connections going through that interface get restarted and if we try to connect to somewhere while the interface is gone, the connection will fail. In the last releases, I have added "checks" that verify that the interface is back before starting any connection, e.g. towards kube-api. If it is not ready, it waits a few seconds and tries again

**Reply by manuelbuil** [reply]
The reason why it works after the restart is because the first time the network gets created. And the second time, it detects that the vxlan network is already there and it does not try creating it again, hence the interface never disappears

**Reply by manuelbuil** [reply]
Try using the latest RKE2 patch versions and you should see improvements in the behaviour. For example, the latest version includes this PR that does the check before contacting kube-api https://github.com/flannel-io/flannel/commit/00014d0926a2b31fc53901d4951c71d69f24f3bf

**Comment by jvnvenu** [comment]
Is flannel support multiple NIC cards? is there any way we can choose NIC cards? I have tried node-ip, the processes are started but kube apis are not reachable, getting http/2 handshake error and no k8s services are reachable

**Reply by manuelbuil** [reply]
Kind of, we should improve it and it's in my backlog. Right now it requires some manual intervention. You need to add to the helm values https://github.com/rancher/rke2-charts/blob/main/charts/rke2-flannel/rke2-flannel/v0.25.400/values.yaml#L11 the flag `--iface=` and then place your nodeIP

**Comment by jvnvenu** [comment]
Even washout changing anything I could see --iface=<ip> parameter in flanneld process arguments. but still kube apis are not reachable, getting http/2 handshake error 

RKE2 Version:
1.30.1

Windows Version:
Windows Server 2019 17763.5329

**Comment by manuelbuil** [comment]
IIUC, you start rke2+flannel in a fresh node (get-hnsnetwork is empty) and kube-api is not reachable. Could you provide the rke2 logs where you see kube-api is not reachable please?

Then you restart the rke2 process, and kube-api is reachable, right?

**Reply by ncoonrod** [reply]
Be sure to run `Get-HnsNetwork` as Administrator. Running it as the signed in user will return empty.

**Comment by ncoonrod** [comment]
Unfortunately, I'm having the same problems described here in v1.28 and v1.30 using Flannel as well. I can't resolve DNS addresses or have my windows pods communicate to any linux pods. I'm using Windows Server 2022.

**Reply by manuelbuil** [reply]
You might be hitting a known Windows issue that started with their July patch release: https://github.com/microsoft/Windows-Containers/issues/516. There is a workaround in the comments related to `FwPerfImprovementChange`, could you give it a try?

**Comment by bsnuggs1** [comment]
I don't know if the problem is the same, but it sounds similar to what I had.  

I had encountered issues with using RKE2 1.29 and 1.30.  

EDIT 2025/05/22 - Noticed a few minutes after posting that there's another discussion about this exact issue in #7440.  You need to upgrade to at least 1.31.7 to have this resolved on the calico side as well.

### TLDR

**Upgrade to 1.31.7.**

The source of the issue is if any HNSEndpoint hits "Destroyed" status, it will cause an index out of bounds panic and cause the process handling felix within RKE2 to crash and the only way to fix is to restart the machine or RKE2.

This issue has been present in RKE2 [1.29](https://github.com/rancher/rke2/blob/release-1.29/go.mod#L6) and [1.30](https://github.com/rancher/rke2/blob/release-1.30/go.mod#L6) since they use [11.7 version of hcsshim](https://github.com/microsoft/hcsshim/blob/v0.11.7/internal/hns/hnsendpoint.go#L32).  This is resolved in [RKE2 1.31](https://github.com/rancher/rke2/blob/release-1.31/go.mod#L6) where its upgraded to [12.8 of hcsshim](https://github.com/microsoft/hcsshim/blob/v0.12.8/internal/hns/hnsendpoint.go#L32) which fixes the bug.

### Long Version

Essentially, what would happen is that DNS resolution within a container would stop functioning randomly for the system I'm running and I was completely lost.  The symptom of this issue can be uncovered if you run the following script in powershell on the node that the container is running on:

```
If((Get-Content {{ C:\var\lib\rancher\rke2\agent\logs\felix.log }} -Last 15 | Select-String "panic: runtime error: index") -eq $null) { Write-Host "no hcsshim crash detected!" -ForegroundColor Green } else { Write-Error "hccshim crash detected!"; exit 1}
```
Assuming that `C:\var\lib\rancher\rke2\agent\logs\felix.log` exists, this will error if a panic runtime is found at the end of the logs.  This means that the process within rke2 has failed.  The full error will look like this:

```
panic: runtime error: index out of range [6] with length 6

goroutine 27 [running]:
github.com/Microsoft/hcsshim/internal/hns.EndpointState.String(0x1f60?)
	/go/pkg/mod/github.com/!microsoft/hcsshim@v0.11.6/internal/hns/hnsendpoint.go:32 +0x38
github.com/projectcalico/calico/felix/dataplane/windows.(*endpointManager).RefreshHnsEndpointCache(0xc00032c3c0, 0x40?)
	/go/src/github.com/projectcalico/calico/felix/dataplane/windows/endpoint_mgr.go:207 +0x8cb
github.com/projectcalico/calico/felix/dataplane/windows.(*endpointManager).CompleteDeferredWork(0xc00032c3c0)
	/go/src/github.com/projectcalico/calico/felix/dataplane/windows/endpoint_mgr.go:328 +0x405
github.com/projectcalico/calico/felix/dataplane/windows.(*WindowsDataplane).apply(0xc0006d2000)
	/go/src/github.com/projectcalico/calico/felix/dataplane/windows/win_dataplane.go:343 +0x63
github.com/projectcalico/calico/felix/dataplane/windows.(*WindowsDataplane).loopUpdatingDataplane(0xc0006d2000)
	/go/src/github.com/projectcalico/calico/felix/dataplane/windows/win_dataplane.go:309 +0x4f0
created by github.com/projectcalico/calico/felix/dataplane/windows.(*WindowsDataplane).Start in goroutine 1
	/go/src/github.com/projectcalico/calico/felix/dataplane/windows/win_dataplane.go:216 +0x56
```

After digging into the source for [hcsshim](https://github.com/microsoft/hcsshim/) and [rke2](https://github.com/rancher/rke2), I was able to find out where the panic came from and it's an issue that has already been resolved in 11.8 of [hcsshim](https://github.com/microsoft/hcsshim/pull/2288) (string method was missing one of the enums).

Turns out that the issue has been present in RKE2 [1.29](https://github.com/rancher/rke2/blob/release-1.29/go.mod#L6) and [1.30](https://github.com/rancher/rke2/blob/release-1.30/go.mod#L6) since they use [11.7 version of hcsshim](https://github.com/microsoft/hcsshim/blob/v0.11.7/internal/hns/hnsendpoint.go#L32).  This is resolved in [RKE2 1.31](https://github.com/rancher/rke2/blob/release-1.31/go.mod#L6) where its upgraded to [12.8 of hcsshim](https://github.com/microsoft/hcsshim/blob/v0.12.8/internal/hns/hnsendpoint.go#L32) which fixes the bug.

![image](https://github.com/user-attachments/assets/7d4e1fa4-f832-4ff4-be7c-3393275d9132)
